[eas-cli] Changes for SSO user type and SSO login.

lzkb opened the pull request:
# Why

These changes are to enable  an SSO user to log in with eas cli via the Expo website.

# How

Added a -sso, -s flag to the login command for a user to select logging in with SSO. The CLI then opens a browser window and redirects the user to the Expo website to log in. The website takes them through the auth flow with their Auth Provider and logs them in to Expo before returning.

This CLI SSO user login follows Expo Go login behavior (which also logs in via the Expo website):
- if no user is logged into the website, the user gets logged in and the session is returned to the CLI
- if a user is already logged into the website then that user session is returned
- once a user is logged into the website, in order to change the logged-in user in the CLI, the current user has to be manually logged out in the website.

Non-SSO user login behavior has not changed.

# Test Plan
Tested manually against the website running locally and with www also running locally. While testing, a couple of log lines were temporarily added to log the session information.

**A) With no user logged in to the website, tested logging in an SSO user:**
eas-cli % EXPO_LOCAL=1 easd login --sso
```
Log in to EAS
Listening for login response at http://localhost:19300
sessionSecret= {"id":"265cd840-2106-46ed-b7f6-6f1835b4e1f1","version":1,"expires_at":2001740400000}
Logged in
```
Website console logs:
```
Navigated to https://dev-49197392.okta.com/oauth2/v1/authorize?client_id=0oa4my0ckz8yKYZum5d7&scope=openid+profile+email+offline_access&response_type=code&redirect_uri=http%3A%2F%2Fexpo.test%2Fauth%2Fcallback%2Fokta&state=6z5w342e3i6y2q2r6i421hs164h1r4ra3b3hk3n3n564w573b5ku1z1i4o1n&nonce=2n3161g1b4h1vj2f6u305m6o4c20k673t5m4k1y2a505i3h3i3r

Navigated to http://expo.test/auth/callback/okta?code=M4XV4Ws6iUAgOwA7vxT3keMVwYOiBskzxiyWpQg1PUk&state=6z5w342e3i6y2q2r6i421hs164h1r4ra3b3hk3n3n564w573b5ku1z1i4o1n

Navigated to http://localhost:19300/auth/callback?username_or_email=liz-sso-viewer&session_secret={%22id%22:%22265cd840-2106-46ed-b7f6-6f1835b4e1f1%22,%22version%22:1,%22expires_at%22:2001740400000}
```

Website and eas-cli have the same user logged in.

**B) With the SSO user already logged in from the test above, tried logging in an SSO user again:**
eas-cli % EXPO_LOCAL=1 easd login --sso
```
Log in to EAS
Listening for login response at http://localhost:19300
sessionSecret= {"id":"265cd840-2106-46ed-b7f6-6f1835b4e1f1","version":1,"expires_at":2001740400000}
Logged in
```
Website immediately returns the session for the logged in user:
```
http://localhost:19300/auth/callback?username_or_email=liz-sso-viewer&session_secret=%7B%22id%22%3A%22265cd840-2106-46ed-b7f6-6f1835b4e1f1%22%2C%22version%22%3A1%2C%22expires_at%22%3A2001740400000%7D
```
Website and eas-cli have the same user logged in.

**C) Checked user identity**
EXPO_LOCAL=1 easd whoami
```
liz-sso-viewer

Accounts:
• liz-sso-viewer (Role: Owner)
• testssoorgforokta (Role: Viewer)
```

**D) Also, with the SSO user logged in, tried logging in a regular user in eas-cli:**
eas-cli % EXPO_LOCAL=1 easd login
```
Log in to EAS
✔ Email or username … lktest
✔ Password … *********
Logged in
```

After this:
eas-cli % EXPO_LOCAL=1 easd whoami

```
lktest

Accounts:
• lktest (Role: Owner)
• lktestorg (Role: Owner)
```

The CLI has a regular user logged in and the website still has the SSO user liz-sso-viewer logged in

It is also possible to log in a regular user with the new flow: 
**E) with nobody logged in to the website, tested logging in a regular user in the website, by clicking on Continue with Password in the website sso-login page:**
eas-cli % EXPO_LOCAL=1 easd login --sso
```
Log in to EAS
Listening for login response at http://localhost:19300
sessionSecret= {"id":"eca11b5d-f4d0-436f-9394-34783ae098a6","version":1,"expires_at":2001740400000}
Logged in
```

eas-cli % EXPO_LOCAL=1 easd whoami 
```
lktest

Accounts:
• lktest (Role: Owner)
• lktestorg (Role: Owner)
lizkaubisch@Lizs-MBP-2 eas-cli % 
```
The regular user is logged in to the CLI and in the website

**F) For latest commit addressing review comments, also tested against staging and tested:**

a) Attempting a command when not logged in, and got new login prompt:
EXPO_LOCAL=1 easd build -p ios --profile preview 

```
An Expo user account is required to proceed.

Use ctrl-c to exit and then run eas login for other login options
Log in to EAS with email or username
? Email or username › 

```

 b) creating an EAS build:
 EXPO_STAGING=1 easd build -p ios --profile preview

```
Compressing project files and uploading to EAS Build. Learn more: https://expo.fyi/eas-build-archive
✔ Uploaded to EAS 

Build details: https://staging.expo.dev/accounts/liz-sso-viewer/projects/liz-sso-viewer-proj1/builds/6e429e23-800b-47d8-b155-73fa9c33ef97

Waiting for build to complete. You can press Ctrl+C to exit.
  Build queued...

Start builds sooner in the priority queue.
Sign up for EAS Production or Enterprise at https://staging.expo.dev/accounts/liz-sso-viewer/settings/subscriptions

Waiting in Free tier queue
|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■| 

⠹ Build in progress...
```

wschurman commented on packages/eas-cli/src/graphql/queries/UserQuery.ts:
```diff
@@ -28,6 +28,10 @@ export const UserQuery = {
                 }
                 ... on SSOUser {
                   username
+                  primaryAccount {
```
can just change the `... on User` to `... on UserActor` and remove this SSOUser spread

wschurman replied:
We may need to add SSOUser to `additionalTypenames` below as well.

lzkb replied:
Done

wschurman commented on packages/eas-cli/src/user/SessionManager.ts:
```diff
@@ -144,13 +145,15 @@ export default class SessionManager {
-  } = {}): Promise<void> {
+  public async showLoginPromptAsync(
+    { nonInteractive = false, printNewLine = false } = {},
+    sso?: boolean | undefined
```
Add to existing named options object rather than a new param

lzkb replied:
Done

wschurman commented on packages/eas-cli/src/user/SessionManager.ts:
```diff
@@ -194,6 +210,16 @@ export default class SessionManager {
+      sessionSecret,
+      userId: id,
+      username,
+      currentConnection: 'Username-Password-Authentication',
```
This value seems incorrect. Not sure what it should be though.

lzkb replied:
Changed to Browser-Flow-Authentication

lzkb replied:
This change also needs to be made in expo cli

wschurman commented on packages/eas-cli/src/user/SessionManager.ts:
```diff
@@ -164,6 +167,19 @@ export default class SessionManager {
 
     Log.log('Log in to EAS');
 
+    // Prompt if sso has not been set (only login will have already set this value)
+    let useSso = sso;
+    if (useSso === undefined) {
+      useSso = await confirmAsync({
+        message: `Do you want to log in with SSO?`,
+      });
+    }
```
I think we probably don't want to add this prompt here since all users will see it (adding friction to workflows). Instead, I think we may want to change the login prompt on L187 to say `Email or username (ctrl-c to exit and then run eas login for other login options):`

lzkb replied:
Done

wschurman commented on packages/eas-cli/src/user/expoSsoLauncher.ts:
```diff
@@ -0,0 +1,82 @@
+}): { executeAuthFlow: () => Promise<string> } => {
+  const { expoWebsiteUrl, serverPort } = options;
+  if (!expoWebsiteUrl || !serverPort) {
+    throw new Error('Expo website URL and localserver port are required.');
```
```suggestion
    throw new Error('Expo website URL and local server port are required.');
```

wschurman commented on packages/eas-cli/src/user/expoSsoLauncher.ts:
```diff
@@ -0,0 +1,82 @@
+  const path = '/auth/callback';
+  const redirectUri = `${scheme}://${hostname}:${serverPort}${path}`;
+
+  const buildExpoSsoLoginUrl = (): string => {
```
optional: no need for this to be a function, especially since it doesn't take any arguments so its return value is constant

lzkb replied:
Left as is

wschurman commented on packages/eas-cli/src/user/expoSsoLauncher.ts:
```diff
@@ -0,0 +1,82 @@
+            }
+            resolve(sessionSecret);
+            response.writeHead(200, { 'Content-Type': 'text/plain' });
+            response.write(`Thank you, website login has completed. You can now close this tab.`);
```
is there a way to close the tab programmatically? Either using JS within the response or maybe a method from the library that provides `openBrowserAsync`?

lzkb replied:
I haven't been able to find a way. The library that opens has nothing for close.

I tried responding with `<script>window.close();</script>` both with Content-Type test/plain and text/html, but this just got displayed.

Will look into further later - https://linear.app/expo/issue/ENG-8924/[eas-cli-expo-cli]-make-clis-close-the-browser-tab-after-user

wschurman commented on packages/eas-cli/src/user/expoSsoLauncher.ts:
```diff
@@ -0,0 +1,82 @@
+export default (options: {
+  expoWebsiteUrl: string;
+  serverPort: number;
+}): { executeAuthFlow: () => Promise<string> } => {
```
prefer to just make this pop the browser rather than return a function that pops the browser (since all callsites just call that function)

```suggestion
}): Promise<string> => {
```

lzkb replied:
Done

wschurman commented on packages/eas-cli/src/user/expoSsoLauncher.ts:
```diff
@@ -0,0 +1,82 @@
+            if (!(request.method === 'GET' && request.url?.includes('/auth/callback'))) {
+              throw new Error('Unexpected SSO login response');
+            }
+            const url = new URL(request.url, `http:${request.headers.host}`);
```
I think the following should work? (so we don't need the odd http thing)

```suggestion
            const url = new URL(request.url);
```

wschurman commented on packages/eas-cli/src/user/expoSsoLauncher.ts:
```diff
@@ -0,0 +1,82 @@
+            const sessionSecret = url.searchParams.get('session_secret');
+
+            if (!sessionSecret) {
+              throw new Error('Login in the website failed unexpectedly.');
```
```suggestion
              throw new Error('Request missing session_secret search param');
```

lzkb replied:
Done

wschurman commented on packages/eas-cli/src/user/fetchUser.ts:
```diff
@@ -0,0 +1,38 @@
+}: {
+  sessionSecret: string;
+}): Promise<{ id: string; username: string }> {
+  const graphqlClient = createGraphqlClient({ accessToken: null, sessionSecret: null });
```
can just pass in `sessionSecret` here?

lzkb replied:
Both args are needed

wschurman replied:
Sorry I wan't clear. I meant can we pass `sessionSecret` here instead of adding it to `fetchOptions` below?

lzkb replied:
I passed it in as recommended and deleted the fetchOptions

wschurman reviewed (commented):
Looks good. Some inline comments and one point of discussion. No major changes though.

wschurman commented on packages/eas-cli/src/user/SessionManager.ts:
```diff
@@ -121,7 +122,7 @@ export default class SessionManager {
     if (!actor) {
       Log.warn('An Expo user account is required to proceed.');
-      await this.showLoginPromptAsync({ nonInteractive, printNewLine: true });
+      await this.showLoginPromptAsync({ nonInteractive, printNewLine: true, sso: undefined });
```
```suggestion
      await this.showLoginPromptAsync({ nonInteractive, printNewLine: true });
```

lzkb replied:
Done

wschurman commented on packages/eas-cli/src/user/SessionManager.ts:
```diff
@@ -147,10 +148,13 @@ export default class SessionManager {
     if (nonInteractive) {
       Errors.error(
-        `Either log in with ${chalk.bold('eas login')} or set the ${chalk.bold(
+        `Either log in with ${chalk.bold('eas login')} or ${chalk.bold(
+          'eas login --sso | -s'
+        )} or set the ${chalk.bold(
```
revert since this in `non-interactive`, and SSO login in impossible in non-interactive.

lzkb replied:
Done

wschurman commented on packages/eas-cli/src/user/SessionManager.ts:
```diff
@@ -162,7 +166,13 @@ export default class SessionManager {
     }
 
-    Log.log('Log in to EAS');
+    if (sso === true) {
```
```suggestion
    if (sso) {
```

lzkb replied:
Done

wschurman commented on packages/eas-cli/src/user/SessionManager.ts:
```diff
@@ -162,7 +166,13 @@ export default class SessionManager {
+      return;
+    }
+
+    Log.log('Use ctrl-c to exit and then run eas login for other login options');
```
```suggestion
    Log.log('For other login options, ctrl-c to exit and then run ${chalk.bold('eas login')}.');
```

lzkb replied:
Done

wschurman commented on packages/eas-cli/src/user/expoSsoLauncher.ts:
```diff
@@ -0,0 +1,80 @@
+
+import Log from '../log';
+
+export async function initiateAuthFlowAsync(options: {
```
```suggestion
export async function getSessionUsingBrowserAuthFlowAsync(options: {
```

lzkb replied:
Done

wschurman commented on packages/eas-cli/src/user/expoSsoLauncher.ts:
```diff
@@ -0,0 +1,80 @@
+            }
+            resolve(sessionSecret);
+            response.writeHead(200, { 'Content-Type': 'text/plain' });
+            response.write(`Thank you, website login has completed. You can now close this tab.`);
```
```suggestion
            response.write(`Website login has completed. You can now close this tab.`);
```

(I think we're trying to reduce the amount of "please" and "thank you" in our messaging)

lzkb replied:
Done

wschurman commented on packages/eas-cli/src/user/expoSsoLauncher.ts:
```diff
@@ -0,0 +1,80 @@
+      );
+
+      server.listen(serverPort, hostname, () => {
+        Log.log('Waiting for website login...');
```
```suggestion
        Log.log('Waiting for browser login...');
```

lzkb replied:
Done

wschurman reviewed (changes requested):
One more round of changes. Checked that it works in my browser and it looks great.

wkozyra95 commented on packages/eas-cli/src/user/expoSsoLauncher.ts:
```diff
@@ -0,0 +1,80 @@
+        }
+      );
+
+      server.listen(serverPort, hostname, () => {
+        Log.log('Waiting for browser login...');
+      });
```
Do you really need to find that port earlier?
If you don't need to know that port before you could do it like this
```suggestion
      server.listen(0, hostname, () => {
        Log.log('Waiting for browser login...');
      });
      const port = server.address().port;
```
(not tested)

lzkb commented on packages/eas-cli/src/api.ts:
```diff
@@ -104,3 +105,14 @@ export function getEASUpdateURL(projectId: string): string {
+export async function getSsoLocalServerPortAsync(): Promise<number> {
+  let startPort: number;
+  if (process.env.SSO_LOCAL_SERVER_PORT) {
+    startPort = parseInt(process.env.SSO_LOCAL_SERVER_PORT, 10);
```
Changed to use Number, thanks

lzkb replied:
Done

szdziedzic commented on packages/eas-cli/README.md:
Did you add these README changes manually?

Just want to mention that we have a script that runs automatically when releasing a new version and generates docs for commands in README, so these changes aren't necessary 😄

lzkb replied:
Oh, thanks, I've reverted the README changes

wkozyra95 replied:
does not seem to be reverted? forgot to push changes?

lzkb commented:
2 changes after discussion with @ide

 1) changed :
 Log.log(`For other login options, ctrl-c to exit and then run ${chalk.bold('eas login')}.`);
 Log.log('Log in to EAS with email or username');

  to:

  Log.log(
     `Log in to EAS with email or username (exit and run ${chalk.bold(
       'eas login'
     )} for other options)`
  );
2) removed the update to CHANGELOG.md,  as SSO is not yet released to  production

wkozyra95 reviewed (approved):
> removed the update to CHANGELOG.md, as SSO is not yet released to production

If you introduced a change it should be in a changelog. If you don't want it to be in changelog, then make the sso option hidden.

szdziedzic commented on packages/eas-cli/src/commands/account/login.ts:
```diff
@@ -1,17 +1,32 @@
 
+  static override flags = {
+    // can pass either --sso or -s
+    sso: Flags.boolean({
+      description: 'Login with SSO',
+      char: 's',
+      default: false,
+    }),
```
```suggestion
    sso: Flags.boolean({
      description: 'Login with SSO',
      char: 's',
      default: false,
      hidden: true
    }),
```

Do the above to make it hidden

lzkb replied:
Done, thanks

ide commented:
We are looking to be able to dogfood SSO but it is not ready for enterprise customers yet. So we need to merge the feature while also not setting the wrong expectation that the feature is available.

wschurman commented:
> We are looking to be able to dogfood SSO but it is not ready for enterprise customers yet. So we need to merge the feature while also not setting the wrong expectation that the feature is available.

IMO it should still be in the changelog, but the entry in the changelog could say something like beta. 
- Changelogs are as much for helping with issue blame and bisection as they are for announcing features.
- I've been adding non-ready features to changelogs forever. Recent example https://github.com/expo/expo/pull/22873 for example.
- Linter complains when there's no changelog entry.

lzkb commented:
> > We are looking to be able to dogfood SSO but it is not ready for enterprise customers yet. So we need to merge the feature while also not setting the wrong expectation that the feature is available.
> 
> IMO it should still be in the changelog, but the entry in the changelog could say something like beta.
> 
> * Changelogs are as much for helping with issue blame and bisection as they are for announcing features.
> * I've been adding non-ready features to changelogs forever. Recent example [[dev-launcher] Add support for SSO users expo#22873](https://github.com/expo/expo/pull/22873). But I could find hundreds more probably (code signing in particular had a ton).
> * Linter complains when there's no changelog entry.



> > We are looking to be able to dogfood SSO but it is not ready for enterprise customers yet. So we need to merge the feature while also not setting the wrong expectation that the feature is available.
> 
> IMO it should still be in the changelog, but the entry in the changelog could say something like beta.
> 
> * Changelogs are as much for helping with issue blame and bisection as they are for announcing features.
> * I've been adding non-ready features to changelogs forever. Recent example [[dev-launcher] Add support for SSO users expo#22873](https://github.com/expo/expo/pull/22873). But I could find hundreds more probably (code signing in particular had a ton).
> * Linter complains when there's no changelog entry.

I've restored the changelog entry
